File: Scala/AoC2024/src/main/scala/Day02.scala
import AoCLib.AoCUtil
import scala.util.boundary, boundary.break

class Day02(day: Int, name: String) extends AoCLib.Solution(day, name):
  override def solve(test: Boolean, index: Int): Unit =
    val input = AoCUtil.readInput(AoCUtil.fileName(day, test))
    val reports = input.map(line => line.split(" ").map( _.toInt))

    val part1SafeCount = solvePartOne(reports)
    solvePartTwo(reports, part1SafeCount)

  def solvePartOne(input: List[Array[Int]]): Int =
    val trends = input.map(reportToTrend)
    val safeTrends = trends.filter(indexOfUnsafeValue(_) < 0)
    val count = safeTrends.size
    println(s"Part One: the total number of safe reports is $count")
    count

  def solvePartTwo(input: List[Array[Int]], part1SafeCount:Int): Unit =
    val trends = input.map(reportToTrend)
    val unsafeTrends = trends.filter(indexOfUnsafeValue(_) >= 0)
    val unsafeMadeSafe = unsafeTrends.filter(canBeMadeSafe)
    val part2SafeCount = unsafeMadeSafe.length
    val total = part1SafeCount + part2SafeCount
    println(s"Part Two: the total number of safe reports is $total")
  
  private def reportToTrend(report:Array[Int]): List[Int] =
    report.toList
      .sliding(2)
      .map(pair => pair(1) - pair.head)
      .toList

  private def mkReport(trend: List[Int]): Array[Int] =
    trend.scan(0)(_ + _).toArray
    
  private def modifyReport(report:Array[Int], indexToRemove:Int): Array[Int] =
    val before = report.take(indexToRemove)
    val after = report.slice(indexToRemove+1, report.length)
    before ++ after
    
  // Returns -1 if all values are safe
  private def indexOfUnsafeValue(trend: List[Int]): Int =
    if trend.head == 0 then return 0
    val isUpward = trend.head > 0
    if isUpward then
      trend.indexWhere(v => { v < 1 || v > 3 })
    else
      trend.indexWhere(v => { v < -3 || v > -1 })

  private def canBeMadeSafe(trend: List[Int]): Boolean =
    val badIndex = indexOfUnsafeValue(trend)
    val report = mkReport(trend)
    val imin = math.max(0, badIndex-1)
    val imax = badIndex+1
    val indexesToRemove = (imin to imax).toList
    indexesToRemove.map(i => reportToTrend(modifyReport(report, i)))
      .map(indexOfUnsafeValue)
      .contains(-1)
    

File: F#/AoC2024/Day02.fs
[<AutoOpen>]
module Day02

open AoC.Util

let parseInput (input:string list) =
    input 
    |> List.map (fun line -> 
        line.Split(" ") 
        |> Array.map (fun s -> int s) 
        |> Array.toList)

let positivePredicate a = a > 0 && a <= 3
let negativePredicate a = a < 0 && a >= -3

let allPositive trend: bool =
    trend |> List.forall positivePredicate

let allNegative trend: bool =
    trend |> List.forall negativePredicate

let isSafe trend: bool =
    (allPositive trend) || (allNegative trend)

let findUnsafeValueIndex (trend: int list) =
    if trend.Head > 0 then
        List.findIndex (positivePredicate >> not) trend
    else
        List.findIndex (negativePredicate >> not) trend

let getTrend report = 
    report
    |> List.windowed 2
    |> List.map (fun [a;b] -> b - a )

let getTrends reports =
    reports
    |> List.map getTrend

let mkReport trend =
    trend |> List.scan (fun acc t -> acc + t) 0 

let solvePartOne (reports: int list list) =
    let trends = getTrends reports
    let safeTrends = trends |> List.filter isSafe
    safeTrends.Length

let solvePartTwo (reports: int list list) pt1SafeCount =
    let trends = getTrends reports
    let unsafeTrends = trends |> List.filter (isSafe >> not)

    let mutable countUnsafeMadeSafe = 0
    for trend in unsafeTrends do
        let mutable canBeMadeSafe = false
        let badIndex = findUnsafeValueIndex trend
        //printfn $"The bad index in {trend} is {badIndex}"
        let report = mkReport trend
        let imin = max 0 (badIndex-1)
        let imax = badIndex+1
        for i in imin .. imax do
            let before = List.take i report
            let after = List.skip (i+1) report
            let modifiedReport = before @ after
            let modifiedTrend = getTrend modifiedReport
            if isSafe modifiedTrend then
                //printfn $"Removing number at {i} made {modifiedReport} safe."
                canBeMadeSafe <- true
        if canBeMadeSafe then
            countUnsafeMadeSafe <- countUnsafeMadeSafe + 1

    pt1SafeCount + countUnsafeMadeSafe

let solveDay02 isTest: Unit =
    let day = 2
    let puzzleName = "Red-Nosed Reports"
    printfn $"Day {day}: {puzzleName}"
    let inputName = inputFileName day isTest
    let input = readInput inputName true

    let reports = parseInput input

    let solution1 = solvePartOne reports
    printfn $"The number of safe reports is {solution1}"

    let solution2 = solvePartTwo reports solution1
    printfn $"The number of safe reports with dampening is {solution2}"

    printfn "All done."
